### PR TITLE
perf(qwen36): tensor core MoE router logits and GQA prefill attention fusion (+23% pp @32k)

### PR DESCRIPTION
**Pull request by fansilas**
## Summary

Two prefill-only kernel levers for the Qwen3.6 MoE path: route the router logits through a bf16 tensor-core GEMM with fp32 output (replacing a per-token warp dot that re-read the whole router weight), and fuse the GQA-8 query heads that share a kv-head so each K page / V tile is loaded once instead of 8x. Prefill-only; decode graph untouched.

## Proof of speedup

- [x] Tested on **RTX 5090** (`sm_120`)

**Decode tok/s** (end-to-end, from `bench/scripts/bench.sh` — fill if this PR targets decode):

| | decode tok/s |
|---|--:|
| before (main) | 411.05 |
| after (this PR) | 410.50 |

**Prefill pp tok/s** (Qwythos / Qwen3.5 — fill if this PR targets prefill; use `--ctx 4096`, `32768`, `65536`, or `131072` and copy the `prefill pp` line — report your best context):

| | prefill pp tok/s |
|---|--:|
| before prefill (main) | 19997.61 |
| after prefill (this PR) | 20897.80 |

Qwythos @32768 (the GQA-attention lever, +4.5%; the router lever is MoE-only). The primary scored model, Qwen3.6-35B-A3B, also gets the router lever: **18803.92 -> 23269.60 pp @32768 (+23.75%)**, median-of-3 A/B +23.5% @32k (16k +19.8%, 4k +11.7%). Prefill fidelity (`qwen3_gguf_prefill_check` @512 int8-KV cont=16): Qwen3.6 top1 1.0000 / KL 0.0074, Qwythos 0.9375 / KL 0.0029.

```text
# bench/scripts/bench.sh @32768  (RTX 5090, sm_120)   before (main)  ->  after (this PR)
# Qwythos-9B Q4_K_M
prefill pp   : 19997.61 tok/s  (ctx=32768, sequential KV fill)  ->  20897.80 tok/s
# Qwen3.6-35B-A3B UD-Q4_K_M
prefill pp   : 18803.92 tok/s  (ctx=32768, sequential KV fill)  ->  23269.60 tok/s
decode tg    : 411.05 tok/s  (n=128, ctx=32768, bs=1)           ->  410.50 tok/s

### CODE REVIEW

**Comment by skyrocket2026** [comment]
## ❌ sparkinfer auto-eval — `8192a79`

| metric | value |
|---|---|
| **label** | `eval:REJECT` |
| Qwen3.5 score | `eval-qwen35:S` (fail) |
| Qwen3.6 score | `eval-qwen36:REJECT` (fail) |
| Qwen3.5 vs same-box main | 17981.29 tok/s → +5.5% (+1037.0) |
| Qwen3.5 scored decode (128 ctx · 128-context) | 254.85 tok/s |
| Qwen3.5 scored prefill (131072 ctx · 128k-context) | 19018.24 pp tok/s · TTFT −5.5% · `eval-prefill:S` |
| Qwen3.5 CB mixed TTFT no-regression gate | 0.494s vs main 0.498s · TTFT −0.8% · pass |
| Qwen3.5 correctness | top-1 97.5% · KL 0.0246 |
| Qwen3.5 128-token no-regression gate | 254.85 tok/s vs main 254.88 tok/s · pass |
| Qwen3.5 4k-context no-regression gate | 246.14 tok/s vs main 246.28 tok/s · pass |
| Qwen3.5 32k-context no-regression gate | 244.53 tok/s vs main 244.59 tok/s · pass |
| Qwen3.5 64k-context no-regression gate | 244.4 tok/s vs main 244.46 tok/s · pass |
| Qwen3.5 128k-context no-regression gate | 243.59 tok/s · pass |
| Qwen3.5 4k prefill no-regression gate | 17992.72 pp tok/s vs main 17687.78 pp tok/s · pass |
| Qwen3.5 32k prefill no-regression gate | 19536.08 pp tok/s vs main 18889.85 pp tok/s · pass |
| Qwen3.5 64k prefill no-regression gate | 19498.13 pp tok/s vs main 18710.01 pp tok/s · pass |
| Qwen3.5 128k prefill no-regression gate | 19018.24 pp tok/s vs main 17981.29 pp tok/s · pass |
| Qwen3.6 vs same-box main | 363.78 tok/s → +0.2% (+0.7) |
| Qwen3.6 scored decode (32768 ctx · 32k-context) | 364.52 tok/s |
| Qwen3.6 scored prefill (32768 ctx · 32k-context) | 20565.48 pp tok/s · TTFT −15.6% · `eval-prefill:L` |
| Qwen3.6 CB mixed TTFT no-regression gate | 0.501s vs main 0.571s · TTFT −12.3% · pass |
| Qwen3.6 correctness | top-1 0.0% · KL 1.4558 |
| Qwen3.6 128-token no-regression gate | 425.77 tok/s vs main 425.74 tok/s · pass |
| Qwen3.6 512-context no-regression gate | 419.38 tok/s vs main 419.57 tok/s · pass |
| Qwen3.6 4k-context no-regression gate | 401.97 tok/s vs main 402.06 tok/s · pass |
| Qwen3.6 16k-context no-regression gate | 388.11 tok/s vs main 387.48 tok/s · pass |
| Qwen3.6 32k-context no-regression gate | 364.52 tok/s vs main 363.78 tok/s · pass |
| Qwen3.6 128 prefill no-regression gate | 0.0 pp tok/s · pass |
| Qwen3.6 512 prefill no-regression gate | 4527.3 pp tok/s vs main 4532.73 pp tok/s · pass |
| Qwen3.6 4k prefill no-regression gate | 14461.72 pp tok/s vs main 13116.78 pp tok/s · pass |
| Qwen3.6 16k prefill no-regression gate | 19637.14 pp tok/s vs main 16772.41 pp tok/s · pass |
| Qwen3.6 32k prefill no-regression gate | 20565.48 pp tok/s vs main 17362.82 pp tok/s · pass |
| **Qwen3.5 optimize** | `eval:S` · 19018.24 tok/s · fail |
| Qwen3.5 optimize — Qwen3.6-35B-A3B guard accuracy | top-1 0.0% · KL 1.4558 · **FAIL** |
| Qwen3.5 optimize — Qwythos-9B (Q4_K_M) 128 | 254.85 tok/s · pass |
| Qwen3.5 optimize — Qwythos-9B (Q4_K_M) 4k | 246.14 tok/s · pass |
| Qwen3.5 optimize — Qwythos-9B (Q4_K_M) 32k | 244.53 tok/s · pass |
| Qwen3.5 optimize — Qwythos-9B (Q4_K_M) 64k | 244.4 tok/s · pass |
| Qwen3.5 optimize — Qwythos-9B (Q4_K_M) 128k | 243.59 tok/s · pass |
| **Qwen3.6 optimize** | `eval:REJECT` · 364.52 tok/s · fail |
| Qwen3.6 optimize — Qwythos-9B (Q4_K_M) guard accuracy | top-1 97.5% · KL 0.0246 · pass |
| Qwen3.6 optimize — Qwen3.6-35B-A3B 128 | 425.77 tok/s · pass |
| Qwen3.6 optimize — Qwen3.6-35B-A3B 512 | 419.38 tok/s · pass |
| Qwen3.6 optimize — Qwen3.6-35B-A3B 4k | 401.97 tok/s · pass |
| Qwen3.6 optimize — Qwen3.6-35B-A3B 16k | 388.11 tok/s · pass |
| Qwen3.6 optimize — Qwen3.6-35B-A3B 32k | 364.52 tok/s · pass |

**Rejected** — correctness gate: top1=0.0 (need >= 0.9), kl=1.455832 (need <= 0.2).

_RTX 5090 (sm_120) · 128/512/4k/16k/32k guarded · Qwen3.5 prefill at 4k/32k/64k/128k · Qwen3.6 prefill at 128/512/4k/16k/32k · scored vs same-box main · built from source · correctness vs llama.cpp. Automated — **not merged**; merge manually after review._

**Comment by skyrocket2026** [comment]
## ❌ sparkinfer auto-eval — `a6b71f0`

| metric | value |
|---|---|
| **label** | `eval:REJECT` |
| Qwen3.5 score | `eval-qwen35:M` (fail) |
| Qwen3.6 score | `eval-qwen36:REJECT` (fail) |
| Qwen3.5 vs same-box main | 17539.21 tok/s → +7.2% (+1364.6) |
| Qwen3.5 scored decode (65536 ctx · 64k-context) | 244.5 tok/s |
| Qwen3.5 scored prefill (131072 ctx · 128k-context) | 18903.82 pp tok/s · TTFT −7.2% · `eval-prefill:M` |
| Qwen3.5 CB mixed TTFT no-regression gate | 0.498s vs main 0.504s · TTFT −1.2% · pass |
| Qwen3.5 correctness | top-1 98.0% · KL 0.0241 |
| Qwen3.5 128-token no-regression gate | 254.82 tok/s vs main 254.92 tok/s · pass |
| Qwen3.5 4k-context no-regression gate | 246.22 tok/s vs main 246.35 tok/s · pass |
| Qwen3.5 32k-context no-regression gate | 244.6 tok/s vs main 244.65 tok/s · pass |
| Qwen3.5 64k-context no-regression gate | 244.5 tok/s vs main 244.52 tok/s · pass |
| Qwen3.5 128k-context no-regression gate | 243.66 tok/s · pass |
| Qwen3.5 4k prefill no-regression gate | 17913.95 pp tok/s vs main 17555.97 pp tok/s · pass |
| Qwen3.5 32k prefill no-regression gate | 19461.68 pp tok/s vs main 18545.91 pp tok/s · pass |
| Qwen3.5 64k prefill no-regression gate | 19403.23 pp tok/s vs main 18210.6 pp tok/s · pass |
| Qwen3.5 128k prefill no-regression gate | 18903.82 pp tok/s vs main 17539.21 pp tok/s · pass |
| Qwen3.6 vs same-box main | 425.89 tok/s → -0.0% (-0.1) |
| Qwen3.6 scored decode (128 ctx · 128-context) | 425.75 tok/s |
| Qwen3.6 scored prefill (32768 ctx · 32k-context) | 20483.14 pp tok/s · TTFT −15.6% · `eval-prefill:L` |
| Qwen3.6 CB mixed TTFT no-regression gate | 0.498s vs main 0.568s · TTFT −12.3% · pass |
| Qwen3.6 correctness | top-1 89.5% · KL 0.0523 |
| Qwen3.6 128-token no-regression gate | 425.75 tok/s vs main 425.89 tok/s · pass |
| Qwen3.6 512-context no-regression gate | 419.2 tok/s vs main 419.42 tok/s · pass |
| Qwen3.6 4k-context no-regression gate | 401.77 tok/s vs main 402.11 tok/s · pass |
| Qwen3.6 16k-context no-regression gate | 387.88 tok/s vs main 388.18 tok/s · pass |
| Qwen3.6 32k-context no-regression gate | 364.2 tok/s vs main 364.5 tok/s · pass |
| Qwen3.6 128 prefill no-regression gate | 0.0 pp tok/s · pass |
| Qwen3.6 512 prefill no-regression gate | 4796.52 pp tok/s vs main 4779.94 pp tok/s · pass |
| Qwen3.6 4k prefill no-regression gate | 14430.1 pp tok/s vs main 12939.52 pp tok/s · pass |
| Qwen3.6 16k prefill no-regression gate | 19561.8 pp tok/s vs main 16683.57 pp tok/s · pass |
| Qwen3.6 32k prefill no-regression gate | 20483.14 pp tok/s vs main 17279.17 pp tok/s · pass |
| **Qwen3.5 optimize** | `eval:M` · 18903.82 tok/s · fail |
| Qwen3.5 optimize — Qwen3.6-35B-A3B guard accuracy | top-1 89.5% · KL 0.0523 · **FAIL** |
| Qwen3.5 optimize — Qwythos-9B (Q4_K_M) 128 | 254.82 tok/s · pass |
| Qwen3.5 optimize — Qwythos-9B (Q4_K_M) 4k | 246.22 tok/s · pass |
| Qwen3.5 optimize — Qwythos-9B (Q4_K_M) 32k | 244.6 tok/s · pass |
| Qwen3.5 optimize — Qwythos-9B (Q4_K_M) 64k | 244.5 tok/s · pass |
| Qwen3.5 optimize — Qwythos-9B (Q4_K_M) 128k | 243.66 tok/s · pass |
| **Qwen3.6 optimize** | `eval:REJECT` · 425.75 tok/s · fail |
| Qwen3.6 optimize — Qwythos-9B (Q4_K_M) guard accuracy | top-1 98.0% · KL 0.0241 · pass |
| Qwen3.6 optimize — Qwen3.6-35B-A3B 128 | 425.75 tok/s · pass |
| Qwen3.6 optimize — Qwen3.6-35B-A3B 512 | 419.2 tok/s · pass |
| Qwen3.6 optimize — Qwen3.6-35B-A3B 4k | 401.77 tok/s · pass |
| Qwen3.6 optimize — Qwen3.6-35B-A3B 16k | 387.88 tok/s · pass |
| Qwen3.6 optimize — Qwen3.6-35B-A3B 32k | 364.2 tok/s · pass |

**Rejected** — correctness gate: top1=0.894942 (need >= 0.9), kl=0.052292 (need <= 0.2).

_RTX 5090 (sm_120) · 128/512/4k/16k/32k guarded · Qwen3.5 prefill at 4k/32k/64k/128k · Qwen3.6 prefill at 128/512/4k/16k/32k · scored vs same-box main · built from source · correctness vs llama.cpp. Automated — **not merged**; merge manually after review._
